// @connectv/html/dist/es5/renderer/plugin/connective/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var pin_1 = require("./pin");
exports.PinPlugin = pin_1.PinPlugin;
var pin_class_1 = require("./pin-class");
exports.PinClassPlugin = pin_class_1.PinClassPlugin;
var object_value_1 = require("./object-value");
exports.PinObjectValuePlugin = object_value_1.PinObjectValuePlugin;
var pin_event_handler_1 = require("./pin-event-handler");
exports.PinEventHandlerPlugin = pin_event_handler_1.PinEventHandlerPlugin;
var input_state_1 = require("./input-state");
exports.InputStatePlugin = input_state_1.InputStatePlugin;
var inner_html_1 = require("./inner-html");
exports.PinInnerHTMLPlugin = inner_html_1.PinInnerHTMLPlugin;
var comp_input_pin_1 = require("./comp-input-pin");
exports.CompInputPinPlugin = comp_input_pin_1.CompInputPinPlugin;
var comp_output_pin_1 = require("./comp-output-pin");
exports.CompOutputPinPlugin = comp_output_pin_1.CompOutputPinPlugin;
var comp_state_bind_1 = require("./comp-state-bind");
exports.CompStateBindPlugin = comp_state_bind_1.CompStateBindPlugin;
var comp_state_track_1 = require("./comp-state-track");
exports.CompStateTrackPlugin = comp_state_track_1.CompStateTrackPlugin;
var comp_output_pin_event_1 = require("./comp-output-pin-event");
exports.CompOutputPinEventsPlugin = comp_output_pin_event_1.CompOutputPinEventsPlugin;
var comp_context_pin_1 = require("./comp-context-pin");
exports.CompContextPinPlugin = comp_context_pin_1.CompContextPinPlugin;
var default_recipient_1 = require("./default-recipient");
exports.DefaultPinLikeRecipientPlugin = default_recipient_1.DefaultPinLikeRecipientPlugin;
function connectivePlugins() {
    return [
        new pin_1.PinPlugin(),
        new pin_class_1.PinClassPlugin(),
        new object_value_1.PinObjectValuePlugin(),
        new inner_html_1.PinInnerHTMLPlugin(),
        new pin_event_handler_1.PinEventHandlerPlugin(),
        new input_state_1.InputStatePlugin(),
        new comp_state_track_1.CompStateTrackPlugin(),
        new comp_state_bind_1.CompStateBindPlugin(),
        new comp_input_pin_1.CompInputPinPlugin(),
        new comp_output_pin_1.CompOutputPinPlugin(),
        new comp_output_pin_event_1.CompOutputPinEventsPlugin(),
        new comp_context_pin_1.CompContextPinPlugin(),
        new default_recipient_1.DefaultPinLikeRecipientPlugin(),
    ];
}
exports.connectivePlugins = connectivePlugins;
var comp_input_pin_2 = require("./comp-input-pin");
exports.CompInputPin = comp_input_pin_2.CompInputPin;
