// @codedoc/core/dist/es6/serve/building-html.js
import { StatusCheckURL, StatusReadyResponse } from './config';
export function buildingHtml(renderer) {
    return renderer.create("html", null,
        renderer.create("head", null,
            renderer.create("link", { href: "https://fonts.googleapis.com/css2?family=Work+Sans:wght@300&display=swap", rel: "stylesheet" }),
            renderer.create("style", null, `
        body {
          font-size: 48px;
          font-family: 'Work Sans', sans-serif;
          background: #eeeeee;
          color: #bdbdbd;
          display: flex;
          justify-content: center;
          align-items: center;
          padding: 0;
          border: 0;
        }
      `)),
        renderer.create("body", null,
            "Building documents, please wait ...",
            renderer.create("script", null, `
        setInterval(function(){
          var xhr = new XMLHttpRequest();
          xhr.open('GET', '${StatusCheckURL}');
          xhr.onload = function() {
            if (xhr.responseText == '${StatusReadyResponse}')
              location.reload();
          };
          xhr.send();
        }, 500);
      `)));
}
