// @connectv/sdh/dist/es5/static/pin/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var html_1 = require("@connectv/html");
var pin_1 = require("./pin");
exports.PinPlugin = pin_1.PinPlugin;
var pin_inner_html_1 = require("./pin-inner-html");
exports.PinInnerHTMLPlugin = pin_inner_html_1.PinInnerHTMLPlugin;
var pin_class_1 = require("./pin-class");
exports.PinClassPlugin = pin_class_1.PinClassPlugin;
var comp_input_pin_1 = require("./comp-input-pin");
exports.CompInputPinPlugin = comp_input_pin_1.CompInputPinPlugin;
var comp_context_pin_1 = require("./comp-context-pin");
exports.CompContextPinPlugin = comp_context_pin_1.CompContextPinPlugin;
function pinPlugins() {
    return [
        new pin_1.PinPlugin(),
        new pin_inner_html_1.PinInnerHTMLPlugin(),
        new pin_class_1.PinClassPlugin(),
        new comp_input_pin_1.CompInputPinPlugin(),
        new comp_context_pin_1.CompContextPinPlugin(),
        new html_1.DefaultPinLikeRecipientPlugin(),
    ];
}
exports.pinPlugins = pinPlugins;
