// @connectv/html/dist/es5/renderer/plugin/rxjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var observable_1 = require("./observable");
exports.ObservablePlugin = observable_1.ObservablePlugin;
var observable_class_1 = require("./observable-class");
exports.ObservableClassPlugin = observable_class_1.ObservableClassPlugin;
var object_value_1 = require("./object-value");
exports.ObservableObjectValuePlugin = object_value_1.ObservableObjectValuePlugin;
var subject_event_handler_1 = require("./subject-event-handler");
exports.SubjectEventHandlerPlugin = subject_event_handler_1.SubjectEventHandlerPlugin;
var input_subject_1 = require("./input-subject");
exports.InputSubjectPlugin = input_subject_1.InputSubjectPlugin;
var inner_html_1 = require("./inner-html");
exports.ObservableInnerHTMLPlugin = inner_html_1.ObservableInnerHTMLPlugin;
var comp_input_subject_1 = require("./comp-input-subject");
exports.CompInputSubjectPlugin = comp_input_subject_1.CompInputSubjectPlugin;
var comp_output_observable_1 = require("./comp-output-observable");
exports.CompOutputObservablePlugin = comp_output_observable_1.CompOutputObservablePlugin;
var comp_output_observable_events_1 = require("./comp-output-observable-events");
exports.CompOutputObservableEventsPlugin = comp_output_observable_events_1.CompOutputObservableEventsPlugin;
var comp_context_subject_1 = require("./comp-context-subject");
exports.CompContextSubjectPlugin = comp_context_subject_1.CompContextSubjectPlugin;
var default_recipient_1 = require("./default-recipient");
exports.DefaultSubjectRecipientPlugin = default_recipient_1.DefaultSubjectRecipientPlugin;
function rxjsPlugins() {
    return [
        new observable_1.ObservablePlugin(),
        new observable_class_1.ObservableClassPlugin(),
        new object_value_1.ObservableObjectValuePlugin(),
        new inner_html_1.ObservableInnerHTMLPlugin(),
        new subject_event_handler_1.SubjectEventHandlerPlugin(),
        new input_subject_1.InputSubjectPlugin(),
        new comp_input_subject_1.CompInputSubjectPlugin(),
        new comp_output_observable_1.CompOutputObservablePlugin(),
        new comp_output_observable_events_1.CompOutputObservableEventsPlugin(),
        new comp_context_subject_1.CompContextSubjectPlugin(),
        new default_recipient_1.DefaultSubjectRecipientPlugin(),
    ];
}
exports.rxjsPlugins = rxjsPlugins;
var comp_input_subject_2 = require("./comp-input-subject");
exports.CompInputSubject = comp_input_subject_2.CompInputSubject;
var trackable_1 = require("./trackable");
exports.trackable = trackable_1.trackable;
exports.Trackable = trackable_1.Trackable;
