// @connectv/html/dist/es6/renderer/plugin/rxjs/index.js
import { ObservablePlugin } from './observable';
import { ObservableClassPlugin } from './observable-class';
import { ObservableObjectValuePlugin } from './object-value';
import { SubjectEventHandlerPlugin } from './subject-event-handler';
import { InputSubjectPlugin } from './input-subject';
import { ObservableInnerHTMLPlugin } from './inner-html';
import { CompInputSubjectPlugin } from './comp-input-subject';
import { CompOutputObservablePlugin } from './comp-output-observable';
import { CompOutputObservableEventsPlugin } from './comp-output-observable-events';
import { CompContextSubjectPlugin } from './comp-context-subject';
import { DefaultSubjectRecipientPlugin } from './default-recipient';
export function rxjsPlugins() {
    return [
        new ObservablePlugin(),
        new ObservableClassPlugin(),
        new ObservableObjectValuePlugin(),
        new ObservableInnerHTMLPlugin(),
        new SubjectEventHandlerPlugin(),
        new InputSubjectPlugin(),
        new CompInputSubjectPlugin(),
        new CompOutputObservablePlugin(),
        new CompOutputObservableEventsPlugin(),
        new CompContextSubjectPlugin(),
        new DefaultSubjectRecipientPlugin(),
    ];
}
export { ObservablePlugin, ObservableClassPlugin, ObservableObjectValuePlugin, ObservableInnerHTMLPlugin, SubjectEventHandlerPlugin, InputSubjectPlugin, CompInputSubjectPlugin, CompContextSubjectPlugin, CompOutputObservablePlugin, CompOutputObservableEventsPlugin, DefaultSubjectRecipientPlugin, };
export { CompInputSubject } from './comp-input-subject';
export { trackable, Trackable } from './trackable';
