// rxline/dist/es5/line/transform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var rxjs_1 = require("rxjs");
var operators_1 = require("rxjs/operators");
var Transform = /** @class */ (function () {
    function Transform(_op$) {
        this.op$ = function (i) {
            try {
                var _o = _op$(i);
                if (_o instanceof Promise)
                    return rxjs_1.from(_o);
                else if (_o instanceof rxjs_1.Observable)
                    return _o;
                else
                    return rxjs_1.of(_o);
            }
            catch (err) {
                return rxjs_1.throwError(err);
            }
        };
    }
    Transform.prototype.combine = function (transform) {
        var _this = this;
        if (transform instanceof Transform)
            return new Transform(function (i) { return _this.op$(i).pipe(operators_1.concatMap(transform.op$)); });
        else
            return this.combine(new Transform(transform));
    };
    Transform.prototype.apply = function (i) { return this.op$(i); };
    return Transform;
}());
exports.Transform = Transform;
function identity() { return new Transform(function (i) { return i; }); }
exports.identity = identity;
function transform(func) { return new Transform(func); }
exports.transform = transform;
