// @connectv/jss-theme/dist/es6/themed-style.js
import jss from 'jss';
import { autoId } from '@connectv/html';
/**
 *
 * Represents a themed style. A themed style can be resolved into
 * JSS stylesheets using a theme object.
 *
 */
export class ThemedStyle {
    /**
     *
     * @param factory the `ThemedStyleFactory` function to be used for resolving this themed style
     *                to JSS stylesheets
     *
     */
    constructor(factory) {
        this.factory = factory;
        this.id = autoId();
    }
    /**
     *
     * @param theme
     * @returns a resolved style object based on given theme
     *
     */
    style(theme) {
        return this.factory(theme);
    }
    /**
     *
     * @param theme
     * @returns a JSS stylesheet based on given theme
     *
     */
    styleSheet(theme) {
        return jss.createStyleSheet(this.style(theme));
    }
}
/**
 *
 * Creates a `ThemedStyle` object from given `ThemedStyleFactory`. The given
 * function will be utilized to resolve the themed style to JSS stylesheets
 * from given theme objects.
 *
 * @param factory
 *
 */
export function themedStyle(factory) {
    return new ThemedStyle(factory);
}
