// @codedoc/core/dist/es5/components/misc/gitter/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@connectv/core");
var renderer_1 = require("../../../transport/renderer");
var holder_1 = require("./holder");
function initGitter(room) {
    if (!window.gitter) {
        var renderer = renderer_1.getRenderer();
        var _state = core_1.state(false);
        var host$ = renderer.create(holder_1.GitterHolder, { state: _state, room: room });
        ((window.gitter = {}).chat = {}).options = {
            room: room,
            activationElement: false,
            targetElement: host$,
        };
        window.gitterState = _state;
        _state.subscribe(function (open) { var _a; return (_a = gitterInstance()) === null || _a === void 0 ? void 0 : _a.toggleChat(open); });
        renderer.render(host$).on(document.body);
        renderer.render(renderer.create("script", { src: "https://sidecar.gitter.im/dist/sidecar.v1.js", async: true, defer: true })).on(document.head);
    }
}
exports.initGitter = initGitter;
function gitterInstance() {
    var _a;
    return (_a = window.gitter.chat) === null || _a === void 0 ? void 0 : _a.defaultChat;
}
exports.gitterInstance = gitterInstance;
function gitterState() {
    if (gitterInstance())
        return window.gitterState;
}
exports.gitterState = gitterState;
