// @codedoc/core/dist/es5/components/darkmode/darklight/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var jss_theme_1 = require("@connectv/jss-theme");
exports.DarkLightStyle = jss_theme_1.themedStyle(function (theme) { return ({
    darklight: {
        position: 'relative',
        overflow: 'hidden',
        'body.dark-mode-animate &>.light, &>.dark': {
            transition: 'opacity .3s, z-index .3s',
        },
        '&>.light': {
            'body.dark &': {
                opacity: 0,
            },
            '@media (prefers-color-scheme: dark)': {
                'body:not(.dark-mode-animate) &': {
                    opacity: 0,
                }
            },
        },
        '&>.dark': {
            position: 'absolute',
            opacity: 0,
            top: 0, left: 0, right: 0,
            zIndex: -1,
            'body.dark &': {
                opacity: 1,
                zIndex: 1,
            },
            '@media (prefers-color-scheme: dark)': {
                'body:not(.dark-mode-animate) &': {
                    opacity: 1,
                    zIndex: 1,
                }
            },
        },
    },
}); });
function InLight(_, renderer, content) {
    return renderer.create("div", { class: "light" }, content);
}
exports.InLight = InLight;
function InDark(_, renderer, content) {
    return renderer.create("div", { class: "dark" }, content);
}
exports.InDark = InDark;
function DarkLight(_, renderer, content) {
    var classes = this.theme.classes(exports.DarkLightStyle);
    return renderer.create("div", { class: classes.darklight }, content);
}
exports.DarkLight = DarkLight;
