// @connectv/jss-theme/dist/es5/themed-style.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var jss_1 = __importDefault(require("jss"));
var html_1 = require("@connectv/html");
/**
 *
 * Represents a themed style. A themed style can be resolved into
 * JSS stylesheets using a theme object.
 *
 */
var ThemedStyle = /** @class */ (function () {
    /**
     *
     * @param factory the `ThemedStyleFactory` function to be used for resolving this themed style
     *                to JSS stylesheets
     *
     */
    function ThemedStyle(factory) {
        this.factory = factory;
        this.id = html_1.autoId();
    }
    /**
     *
     * @param theme
     * @returns a resolved style object based on given theme
     *
     */
    ThemedStyle.prototype.style = function (theme) {
        return this.factory(theme);
    };
    /**
     *
     * @param theme
     * @returns a JSS stylesheet based on given theme
     *
     */
    ThemedStyle.prototype.styleSheet = function (theme) {
        return jss_1.default.createStyleSheet(this.style(theme));
    };
    return ThemedStyle;
}());
exports.ThemedStyle = ThemedStyle;
/**
 *
 * Creates a `ThemedStyle` object from given `ThemedStyleFactory`. The given
 * function will be utilized to resolve the themed style to JSS stylesheets
 * from given theme objects.
 *
 * @param factory
 *
 */
function themedStyle(factory) {
    return new ThemedStyle(factory);
}
exports.themedStyle = themedStyle;
