// @connectv/html/dist/es5/renderer/plugin/component/basic-plugins.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var plugin_1 = require("../plugin");
function isCompProcessPlugin(whatever) {
    return whatever && whatever.prepare && typeof whatever.prepare == 'function';
}
exports.isCompProcessPlugin = isCompProcessPlugin;
function isCompIOPlugin(whatever) {
    return whatever && whatever.wire && typeof whatever.wire == 'function';
}
exports.isCompIOPlugin = isCompIOPlugin;
function isCompPropPlugin(whatever) {
    return whatever && whatever.wireProp && typeof whatever.wireProp == 'function';
}
exports.isCompPropPlugin = isCompPropPlugin;
function isCompContextPlugin(whatever) {
    return whatever && whatever.wireContext && typeof whatever.wireContext == 'function';
}
exports.isCompContextPlugin = isCompContextPlugin;
var DefaultReactiveRecipientPlugin = /** @class */ (function () {
    function DefaultReactiveRecipientPlugin() {
        this.priority = plugin_1.PluginPriority.Fallback;
    }
    DefaultReactiveRecipientPlugin.prototype.unique = function (plugin) {
        return plugin instanceof DefaultReactiveRecipientPlugin;
    };
    return DefaultReactiveRecipientPlugin;
}());
exports.DefaultReactiveRecipientPlugin = DefaultReactiveRecipientPlugin;
function isDefaultReactiveRecipientPlugin(whatever) {
    return whatever instanceof DefaultReactiveRecipientPlugin;
}
exports.isDefaultReactiveRecipientPlugin = isDefaultReactiveRecipientPlugin;
