// @connectv/html/dist/es5/renderer/plugin/component/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@connectv/core");
var Component = /** @class */ (function () {
    function Component(props, children, _adapter) {
        this.props = props;
        this.children = children;
        this._adapter = _adapter;
        _adapter.track(this);
        this.init();
        var _signature = this.signature();
        if (_signature)
            _adapter.expose(_signature);
    }
    Component.prototype.init = function () { };
    Component.prototype.bind = function () { };
    Component.prototype.clear = function () { };
    Component.prototype.signature = function () { return undefined; };
    Component.prototype.track = function (sub) {
        if (core_1.isBindable(sub) || core_1.isClearable(sub))
            this._adapter.track(sub);
        else
            this._adapter.track({
                clear: function () { sub.unsubscribe(); }
            });
    };
    Component.__CVH_component_class__ = true;
    return Component;
}());
exports.Component = Component;
function isCompClass(comp) {
    return comp.__CVH_component_class__;
}
exports.isCompClass = isCompClass;
function isCompInputWithOptions(whatever) {
    return whatever && whatever.inputOptions;
}
exports.isCompInputWithOptions = isCompInputWithOptions;
