// @connectv/sdh/dist/es5/static/observable/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var observable_1 = require("./observable");
exports.ObservablePlugin = observable_1.ObservablePlugin;
var observable_inner_html_1 = require("./observable-inner-html");
exports.ObservableInnerHTMLPlugin = observable_inner_html_1.ObservableInnerHTMLPlugin;
var observable_class_1 = require("./observable-class");
exports.ObservableClassPlugin = observable_class_1.ObservableClassPlugin;
var comp_input_subject_1 = require("./comp-input-subject");
exports.CompInputSubjectPlugin = comp_input_subject_1.CompInputSubjectPlugin;
var comp_context_subject_1 = require("./comp-context-subject");
exports.CompContextSubjectPlugin = comp_context_subject_1.CompContextSubjectPlugin;
function observablePlugins() {
    return [
        new observable_1.ObservablePlugin(),
        new observable_inner_html_1.ObservableInnerHTMLPlugin(),
        new observable_class_1.ObservableClassPlugin(),
        new comp_input_subject_1.CompInputSubjectPlugin(),
        new comp_context_subject_1.CompContextSubjectPlugin(),
    ];
}
exports.observablePlugins = observablePlugins;
