// @connectv/html/dist/es6/renderer/plugin/connective/index.js
import { PinPlugin } from './pin';
import { PinClassPlugin } from './pin-class';
import { PinObjectValuePlugin } from './object-value';
import { PinEventHandlerPlugin } from './pin-event-handler';
import { InputStatePlugin } from './input-state';
import { PinInnerHTMLPlugin } from './inner-html';
import { CompInputPinPlugin } from './comp-input-pin';
import { CompOutputPinPlugin } from './comp-output-pin';
import { CompStateBindPlugin } from './comp-state-bind';
import { CompStateTrackPlugin } from './comp-state-track';
import { CompOutputPinEventsPlugin } from './comp-output-pin-event';
import { CompContextPinPlugin } from './comp-context-pin';
import { DefaultPinLikeRecipientPlugin } from './default-recipient';
export function connectivePlugins() {
    return [
        new PinPlugin(),
        new PinClassPlugin(),
        new PinObjectValuePlugin(),
        new PinInnerHTMLPlugin(),
        new PinEventHandlerPlugin(),
        new InputStatePlugin(),
        new CompStateTrackPlugin(),
        new CompStateBindPlugin(),
        new CompInputPinPlugin(),
        new CompOutputPinPlugin(),
        new CompOutputPinEventsPlugin(),
        new CompContextPinPlugin(),
        new DefaultPinLikeRecipientPlugin(),
    ];
}
export { PinPlugin, PinClassPlugin, PinObjectValuePlugin, PinInnerHTMLPlugin, PinEventHandlerPlugin, InputStatePlugin, CompInputPinPlugin, CompContextPinPlugin, CompStateBindPlugin, CompStateTrackPlugin, CompOutputPinPlugin, CompOutputPinEventsPlugin, DefaultPinLikeRecipientPlugin, };
export { CompInputPin } from './comp-input-pin';
