// @connectv/html/dist/es5/renderer/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var extensible_renderer_1 = require("./extensible-renderer");
exports.ExtensibleRenderer = extensible_renderer_1.ExtensibleRenderer;
var life_cycle_1 = require("./plugin/life-cycle");
var event_handler_1 = require("./plugin/event-handler");
var ref_1 = require("./plugin/ref");
var object_value_1 = require("./plugin/object-value");
var inner_html_1 = require("./plugin/inner-html");
var component_1 = require("./plugin/component/component");
var track_1 = require("./plugin/component/track");
var expose_1 = require("./plugin/component/expose");
var rxjs_1 = require("./plugin/rxjs");
var connective_1 = require("./plugin/connective");
var types_1 = require("./plugin/component/types");
var check_inputs_1 = require("./plugin/component/check-inputs");
var state_io_1 = require("./plugin/component/state-io");
var context_1 = require("./plugin/component/context");
var ConnectiveRenderer = /** @class */ (function (_super) {
    __extends(ConnectiveRenderer, _super);
    function ConnectiveRenderer() {
        return _super.apply(this, __spreadArrays([new life_cycle_1.LifeCyclePlugin(),
            new event_handler_1.EventHandlerPlugin(),
            new ref_1.RefPlugin(),
            new object_value_1.ObjectValuePlugin(),
            new inner_html_1.InnerHTMLPlugin(),
            new component_1.ComponentPlugin(),
            new track_1.TrackPlugin(),
            new expose_1.ExposePlugin(),
            new context_1.ContextPlugin(),
            new state_io_1.CompStateIOPlugin(),
            new check_inputs_1.CheckCompInputsPlugin()], rxjs_1.rxjsPlugins(), connective_1.connectivePlugins())) || this;
    }
    return ConnectiveRenderer;
}(extensible_renderer_1.ExtensibleRenderer));
exports.ConnectiveRenderer = ConnectiveRenderer;
exports.default = ConnectiveRenderer;
var Component = /** @class */ (function (_super) {
    __extends(Component, _super);
    function Component() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return Component;
}(types_1.Component));
exports.Component = Component;
var renderer_1 = require("./renderer");
exports.RawRenderer = renderer_1.Renderer;
var ref_2 = require("./ref");
exports.Ref = ref_2.Ref;
exports.ref = ref_2.ref;
__export(require("./plugin"));
__export(require("./error"));
