// @connectv/sdh/dist/es5/static/promise/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var promise_1 = require("./promise");
exports.PromisePlugin = promise_1.PromisePlugin;
var promise_inner_html_1 = require("./promise-inner-html");
exports.PromiseInnerHTMLPlugin = promise_inner_html_1.PromiseInnerHTMLPlugin;
var promise_class_1 = require("./promise-class");
exports.PromiseClassPlugin = promise_class_1.PromiseClassPlugin;
var comp_input_promise_1 = require("./comp-input-promise");
exports.CompInputPromisePlugin = comp_input_promise_1.CompInputPromisePlugin;
var default_recipient_1 = require("./default-recipient");
exports.DefaultPromiseRecipientPlugin = default_recipient_1.DefaultPromiseRecipientPlugin;
var comp_context_promise_1 = require("./comp-context-promise");
exports.CompContextPromisePlugin = comp_context_promise_1.CompContextPromisePlugin;
function promisePlugins() {
    return [
        new promise_1.PromisePlugin(),
        new promise_inner_html_1.PromiseInnerHTMLPlugin(),
        new promise_class_1.PromiseClassPlugin(),
        new comp_input_promise_1.CompInputPromisePlugin(),
        new comp_context_promise_1.CompContextPromisePlugin(),
        new default_recipient_1.DefaultPromiseRecipientPlugin(),
    ];
}
exports.promisePlugins = promisePlugins;
var recipient_promise_1 = require("./recipient-promise");
exports.isRecipientPromise = recipient_promise_1.isRecipientPromise;
exports.recipientPromise = recipient_promise_1.recipientPromise;
