// @connectv/html/dist/es5/renderer/plugin/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./basic-plugins"));
var plugin_1 = require("./plugin");
exports.PluginPriority = plugin_1.PluginPriority;
var event_handler_1 = require("./event-handler");
exports.EventHandlerPlugin = event_handler_1.EventHandlerPlugin;
var inner_html_1 = require("./inner-html");
exports.InnerHTMLPlugin = inner_html_1.InnerHTMLPlugin;
var life_cycle_1 = require("./life-cycle");
exports.LifeCyclePlugin = life_cycle_1.LifeCyclePlugin;
var object_value_1 = require("./object-value");
exports.ObjectValuePlugin = object_value_1.ObjectValuePlugin;
var ref_1 = require("./ref");
exports.RefPlugin = ref_1.RefPlugin;
var styled_1 = require("./styled");
exports.StyledPlugin = styled_1.StyledPlugin;
exports.styled = styled_1.styled;
__export(require("./rxjs"));
__export(require("./connective"));
__export(require("./component"));
