// @connectv/html/dist/es5/renderer/plugin/styled.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var plugin_1 = require("./plugin");
var _ClassListFixed = __importStar(require("../../shared/class-list-fixed"));
var StyledPlugin = /** @class */ (function () {
    function StyledPlugin(map) {
        this.map = map;
        this.priority = plugin_1.PluginPriority.High;
    }
    StyledPlugin.prototype.unique = function (plugin) {
        return plugin instanceof StyledPlugin;
    };
    StyledPlugin.prototype.postCreate = function (node) {
        if (node instanceof HTMLElement) {
            var tag = node.tagName.toLocaleLowerCase();
            if (tag in this.map) {
                var clazz = this.map[tag];
                node.classList.add(clazz);
                _ClassListFixed.add(node, clazz);
            }
        }
    };
    return StyledPlugin;
}());
exports.StyledPlugin = StyledPlugin;
function styled(map) {
    return new StyledPlugin(map);
}
exports.styled = styled;
exports.default = styled;
