// @codedoc/core/dist/es5/config/defaults.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var marked_1 = require("@connectv/marked");
var theme_1 = require("../theme");
var guess_title_1 = require("../transport/guess-title");
var code_1 = require("../components/code");
var heading_1 = require("../components/heading");
var button_1 = require("../components/button");
var tabs_1 = require("../components/tabs");
var collapse_1 = require("../components/collapse");
var meta_override_1 = require("../components/page/meta-override");
var darklight_1 = require("../components/darkmode/darklight");
var github_1 = require("../components/misc/github");
var watermark_1 = require("../components/misc/watermark");
var icon_1 = require("../components/misc/icon");
var selection_1 = require("../components/code/selection");
var same_line_length_1 = require("../components/code/same-line-length");
var line_hint_1 = require("../components/code/line-hint");
var line_ref_1 = require("../components/code/line-ref");
var smart_copy_1 = require("../components/code/smart-copy");
var copy_headings_1 = require("../components/heading/copy-headings");
var highlight_1 = require("../components/page/contentnav/highlight");
var deferred_iframe_1 = require("../transport/deferred-iframe");
var heading_2 = require("../components/page/toc/heading");
var smooth_loading_1 = require("../transport/smooth-loading");
var toc_highlight_1 = require("../components/page/toc/toc-highlight");
var prevnext_1 = require("../components/page/toc/prevnext");
var post_nav_1 = require("../components/page/toc/search/post-nav");
var add_line_ids_1 = require("../components/code/line-links/add-line-ids");
var copy_line_link_1 = require("../components/code/line-links/copy-line-link");
exports.DefaultMarkdownCustomComponents = {
    Tab: tabs_1.Tab, Tabs: tabs_1.Tabs, Collapse: collapse_1.Collapse,
    Button: button_1.Button, Buttons: button_1.Buttons, CopyButton: button_1.CopyButton,
    DarkLight: darklight_1.DarkLight, InDark: darklight_1.InDark, InLight: darklight_1.InLight,
    GithubButton: github_1.GithubButton, Watermark: watermark_1.Watermark,
    MetaOverride: meta_override_1.MetaOverride,
    ToCPrevNext: prevnext_1.ToCPrevNext$,
};
exports.DefaultMarkdownCustomInlineComponents = {
    Icon: icon_1.Icon,
};
exports.DefaultToCMarkdownCustomComponents = {
    Button: button_1.Button, Buttons: button_1.Buttons, Collapse: collapse_1.Collapse,
    DarkLight: darklight_1.DarkLight, InDark: darklight_1.InDark, InLight: darklight_1.InLight,
    GithubButton: github_1.GithubButton, Watermark: watermark_1.Watermark,
};
exports.DefaultToCMarkdownCustomInlineComponents = {
    Icon: icon_1.Icon,
};
exports.DefaultConfig = {
    src: {
        base: 'docs/md',
        toc: '_toc.md',
        pick: /\.md$/,
        drop: /(^_)|(\/_)/,
        not_found: '404.md',
    },
    dest: {
        html: '.',
        assets: '.',
        bundle: 'docs/assets',
        styles: 'docs/assets',
        namespace: '',
    },
    bundle: {
        init: [
            selection_1.codeSelection$, same_line_length_1.sameLineLengthInCodes$, line_hint_1.codeLineHints$, line_ref_1.codeLineRef$, smart_copy_1.smartCopy$,
            copy_headings_1.copyHeadings$, highlight_1.contentNavHighlight$, deferred_iframe_1.deferredIframes$, smooth_loading_1.smoothLoading$, toc_highlight_1.tocHighlight$,
            post_nav_1.postNavSearch$, copy_line_link_1.copyLineLiks$,
        ],
    },
    dev: {
        port: 3000
    },
    page: {
        title: {
            base: 'New Codedoc Project',
            connector: ' | ',
            extractor: function (content, config) { return guess_title_1.guessTitle(content, config.page.title.base, config.page.title.connector); },
        },
        post: [add_line_ids_1.addLineIds]
    },
    afterBuild: [],
    theme: theme_1.DefaultTheme,
    markdown: {
        Code: code_1.Code,
        Heading: heading_1.Heading,
        customComponents: exports.DefaultMarkdownCustomComponents,
        customInlineComponents: exports.DefaultMarkdownCustomInlineComponents,
        BlockQuote: marked_1.quotedComponents(exports.DefaultMarkdownCustomComponents),
        Link: marked_1.linkedComponents(exports.DefaultMarkdownCustomInlineComponents),
    },
    tocMarkdown: {
        Heading: heading_2.ToCHeading,
        customComponents: exports.DefaultToCMarkdownCustomComponents,
        customInlineComponents: exports.DefaultToCMarkdownCustomInlineComponents,
        BlockQuote: marked_1.quotedComponents(exports.DefaultToCMarkdownCustomComponents),
        Link: marked_1.linkedComponents(exports.DefaultToCMarkdownCustomInlineComponents),
    },
};
